### DISCUSSION TITLE
Specifying experimental data in a separate file

### DISCUSSION THREAD

**Discussion by k-arakane** [biomass-dev/biomass#274] [Ideas]
Currently, experimental data that is used for the optimization of the model parameters need to be specified within the `observable.py` under the model directory. However, when exploring various model structures (especially when using Text2Model), copy-and-pasting this information from one model directory to another can become a tedious task and is also a potential source of errors. I propose that being able to define experimental data in a separate file can solve these problems by allowing the user to simply copy the file into the newly created model directory.

My current ideas on how to implement this feature are something like the following:

- The file should have a predefined name (e.g., `exp_data.txt`), and biomass will detect if a file with this specific name is present within the model directory when a model object is being instantiated
  - When the file is present, the functions that involve experimental data such as `set_data` and `get_timepoints` will return data that was read from this file; otherwise, it will behave the same as before
  - Users who wish to can specify the experimental data in `observable.py` like before, maintaining backwards compatibility
- All aspects of the experimental data (the data values, time points, and variability values) should be specifiable within this file
- The format will be based on YAML (but not entirely the same), which is rather human-friendly and has a Text2Model-esque feel
  - This can also be helpful since we can take advantage of syntax highlighting for YAML files

To give you an idea of how the actual file might look, I have attached several example files:

- [example_1.txt](https://github.com/user-attachments/files/15791863/example_1.txt)
- [example_2.txt](https://github.com/user-attachments/files/15791865/example_2.txt)
- [example_3.txt](https://github.com/user-attachments/files/15791867/example_3.txt)

I am currently working on developing parsers using these examples, and I would be very happy to know what you think!

**Comment by himoto** [comment]
Sounds exciting! Comments:
- This new feature would definitely help with data entry. As you mentioned, currently experimental data must be specified in `observable.py` file. I would like to see how the [Tutorial 1](https://biomass-core.readthedocs.io/en/latest/tutorial/cfos.html#model-preparation) changes after this idea is implemented. When you send out a PR on this, can you please work on the docs as well? Overall, I agree with the idea of having a separate file for the data.
- New test code is needed for this feature. You can use the Nakakuki model, the Kholodenko1999 model, or any toy models you have made.

**Reply by k-arakane** [reply]
Thanks for commenting @himoto!
I would definitely like to address the points raised here. I will also try my best to match the amount of comments/docstrings within the rest of the code to aid the review process.
